src/middlewared/middlewared/plugins/virt/instance.py

```diff
@@ -338,3 +339,24 @@ async def restart(self, job, id, data):
         }})
 
         return True
+
+    @private
+    def get_shell(self, id):
+        """
+        Method to get a valid shell to be used by default.
+        """
+
+        self.middleware.call_sync('virt.global.check_initialized')
+        instance = self.middleware.call_sync('virt.instance.get_instance', id)
+        if instance['type'] != 'CONTAINER':
+            raise CallError('Only available for containers.')
+        if instance['status'] != 'RUNNING':
+            raise CallError('Container must be running.')
+        config = self.middleware.call_sync('virt.global.config')
+        mount_info = self.middleware.call_sync('filesystem.mount_info', [['mount_source', '=', f'{config["dataset"]}/containers/{id}']])
+        if not mount_info:
+            return None
```

Comment:
Do you think we should maybe error out if we are not able to identify a shell for an instance as otherwise this would end up returning none which then will be used in the subprocess

Reply:
How about handling it on the caller, defaulting to /bin/sh?

Reply:
I like that approach ✅